File: Languages/npeg_c++/robusthaven.tests/ParserTest_MathematicalFormula/MathematicalFormula.cpp
#include "MathematicalFormula.h"

using namespace RobustHaven::Text;

int MathematicalFormula::MathematicalFormula_impl_0()
{
	return this->MathematicalFormula_impl_1();
}
int MathematicalFormula::MathematicalFormula_impl_1()
{
	const char* _nodeName_0 = "EXPRESSION";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_2, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_2()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_3, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_28);
}
int MathematicalFormula::MathematicalFormula_impl_28()
{
	return this->zeroOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_29);
}
int MathematicalFormula::MathematicalFormula_impl_29()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_30, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_34);
}
int MathematicalFormula::MathematicalFormula_impl_34()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_35, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_44);
}
int MathematicalFormula::MathematicalFormula_impl_44()
{
	return this->zeroOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_45);
}
int MathematicalFormula::MathematicalFormula_impl_45()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_46, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_50);
}
int MathematicalFormula::MathematicalFormula_impl_50()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_51, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_54);
}
int MathematicalFormula::MathematicalFormula_impl_54()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_55, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_58);
}
int MathematicalFormula::MathematicalFormula_impl_58()
{
	const char* _literal_0 = ")";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_55()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_56, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_57);
}
int MathematicalFormula::MathematicalFormula_impl_57()
{
	return this->recursionCall((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_0);
}
int MathematicalFormula::MathematicalFormula_impl_56()
{
	const char* _literal_0 = "(";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_51()
{
	const char* _nodeName_0 = "VALUE";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_52, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_52()
{
	return this->oneOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_53);
}
int MathematicalFormula::MathematicalFormula_impl_53()
{
	const char* _classExpression_0 = "[0-9]";

	return this->characterClass(_classExpression_0, 5);
}
int MathematicalFormula::MathematicalFormula_impl_46()
{
	const char* _nodeName_0 = "SYMBOL";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_47, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_47()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_48, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_49);
}
int MathematicalFormula::MathematicalFormula_impl_49()
{
	const char* _literal_0 = "/";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_48()
{
	const char* _literal_0 = "*";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_35()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_36, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_39);
}
int MathematicalFormula::MathematicalFormula_impl_39()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_40, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_43);
}
int MathematicalFormula::MathematicalFormula_impl_43()
{
	const char* _literal_0 = ")";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_40()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_41, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_42);
}
int MathematicalFormula::MathematicalFormula_impl_42()
{
	return this->recursionCall((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_0);
}
int MathematicalFormula::MathematicalFormula_impl_41()
{
	const char* _literal_0 = "(";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_36()
{
	const char* _nodeName_0 = "VALUE";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_37, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_37()
{
	return this->oneOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_38);
}
int MathematicalFormula::MathematicalFormula_impl_38()
{
	const char* _classExpression_0 = "[0-9]";

	return this->characterClass(_classExpression_0, 5);
}
int MathematicalFormula::MathematicalFormula_impl_30()
{
	const char* _nodeName_0 = "SYMBOL";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_31, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_31()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_32, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_33);
}
int MathematicalFormula::MathematicalFormula_impl_33()
{
	const char* _literal_0 = "-";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_32()
{
	const char* _literal_0 = "+";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_3()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_4, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_13);
}
int MathematicalFormula::MathematicalFormula_impl_13()
{
	return this->zeroOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_14);
}
int MathematicalFormula::MathematicalFormula_impl_14()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_15, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_19);
}
int MathematicalFormula::MathematicalFormula_impl_19()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_20, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_23);
}
int MathematicalFormula::MathematicalFormula_impl_23()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_24, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_27);
}
int MathematicalFormula::MathematicalFormula_impl_27()
{
	const char* _literal_0 = ")";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_24()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_25, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_26);
}
int MathematicalFormula::MathematicalFormula_impl_26()
{
	return this->recursionCall((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_0);
}
int MathematicalFormula::MathematicalFormula_impl_25()
{
	const char* _literal_0 = "(";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_20()
{
	const char* _nodeName_0 = "VALUE";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_21, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_21()
{
	return this->oneOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_22);
}
int MathematicalFormula::MathematicalFormula_impl_22()
{
	const char* _classExpression_0 = "[0-9]";

	return this->characterClass(_classExpression_0, 5);
}
int MathematicalFormula::MathematicalFormula_impl_15()
{
	const char* _nodeName_0 = "SYMBOL";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_16, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_16()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_17, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_18);
}
int MathematicalFormula::MathematicalFormula_impl_18()
{
	const char* _literal_0 = "/";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_17()
{
	const char* _literal_0 = "*";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_4()
{
	return this->prioritizedChoice((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_5, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_8);
}
int MathematicalFormula::MathematicalFormula_impl_8()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_9, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_12);
}
int MathematicalFormula::MathematicalFormula_impl_12()
{
	const char* _literal_0 = ")";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_9()
{
	return this->sequence((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_10, (Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_11);
}
int MathematicalFormula::MathematicalFormula_impl_11()
{
	return this->recursionCall((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_0);
}
int MathematicalFormula::MathematicalFormula_impl_10()
{
	const char* _literal_0 = "(";

	return this->literal(_literal_0, 1, 1);
}
int MathematicalFormula::MathematicalFormula_impl_5()
{
	const char* _nodeName_0 = "VALUE";

	return this->capturingGroup((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_6, _nodeName_0, 0, NULL);
}
int MathematicalFormula::MathematicalFormula_impl_6()
{
	return this->oneOrMore((Npeg::IsMatchPredicate)&MathematicalFormula::MathematicalFormula_impl_7);
}
int MathematicalFormula::MathematicalFormula_impl_7()
{
	const char* _classExpression_0 = "[0-9]";

	return this->characterClass(_classExpression_0, 5);
}

File: Languages/npeg_csharp/RobustHaven.Text.Npeg.Tests/Parsers/MathematicalFormula.cs
using System;

namespace RobustHaven.Text.Npeg.Tests.Parsers
{
	public class MathematicalFormula : NpegParser
	{
		public MathematicalFormula(InputIterator iterator) : base(iterator)
		{
		}

		public override Boolean IsMatch()
		{
			return MathematicalFormula_impl_0();
		}


		protected Boolean MathematicalFormula_impl_0()
		{
			return MathematicalFormula_impl_1();
		}

		protected Boolean MathematicalFormula_impl_1()
		{
			String _nodeName_0 = "Expr";
			return CapturingGroup(MathematicalFormula_impl_2, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_2()
		{
			String _nodeName_0 = "Sum";
			return CapturingGroup(MathematicalFormula_impl_3, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_3()
		{
			return Sequence(MathematicalFormula_impl_4, MathematicalFormula_impl_30);
		}

		protected Boolean MathematicalFormula_impl_30()
		{
			return ZeroOrMore(MathematicalFormula_impl_31, "");
		}

		protected Boolean MathematicalFormula_impl_31()
		{
			return Sequence(MathematicalFormula_impl_32, MathematicalFormula_impl_36);
		}

		protected Boolean MathematicalFormula_impl_36()
		{
			String _nodeName_0 = "Product";
			return CapturingGroup(MathematicalFormula_impl_37, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_37()
		{
			return Sequence(MathematicalFormula_impl_38, MathematicalFormula_impl_47);
		}

		protected Boolean MathematicalFormula_impl_47()
		{
			return ZeroOrMore(MathematicalFormula_impl_48, "");
		}

		protected Boolean MathematicalFormula_impl_48()
		{
			return Sequence(MathematicalFormula_impl_49, MathematicalFormula_impl_53);
		}

		protected Boolean MathematicalFormula_impl_53()
		{
			String _nodeName_0 = "Value";
			return CapturingGroup(MathematicalFormula_impl_54, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_54()
		{
			return PrioritizedChoice(MathematicalFormula_impl_55, MathematicalFormula_impl_57);
		}

		protected Boolean MathematicalFormula_impl_57()
		{
			return Sequence(MathematicalFormula_impl_58, MathematicalFormula_impl_61);
		}

		protected Boolean MathematicalFormula_impl_61()
		{
			String _literal_0 = ")";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_58()
		{
			return Sequence(MathematicalFormula_impl_59, MathematicalFormula_impl_60);
		}

		protected Boolean MathematicalFormula_impl_60()
		{
			return RecursionCall(MathematicalFormula_impl_0);
		}

		protected Boolean MathematicalFormula_impl_59()
		{
			String _literal_0 = "(";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_55()
		{
			return OneOrMore(MathematicalFormula_impl_56, "");
		}

		protected Boolean MathematicalFormula_impl_56()
		{
			String _classExpression_0 = "[0-9]";
			return CharacterClass(_classExpression_0, 5);
		}

		protected Boolean MathematicalFormula_impl_49()
		{
			String _nodeName_0 = "Symbol";
			return CapturingGroup(MathematicalFormula_impl_50, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_50()
		{
			return PrioritizedChoice(MathematicalFormula_impl_51, MathematicalFormula_impl_52);
		}

		protected Boolean MathematicalFormula_impl_52()
		{
			String _literal_0 = "/";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_51()
		{
			String _literal_0 = "*";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_38()
		{
			String _nodeName_0 = "Value";
			return CapturingGroup(MathematicalFormula_impl_39, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_39()
		{
			return PrioritizedChoice(MathematicalFormula_impl_40, MathematicalFormula_impl_42);
		}

		protected Boolean MathematicalFormula_impl_42()
		{
			return Sequence(MathematicalFormula_impl_43, MathematicalFormula_impl_46);
		}

		protected Boolean MathematicalFormula_impl_46()
		{
			String _literal_0 = ")";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_43()
		{
			return Sequence(MathematicalFormula_impl_44, MathematicalFormula_impl_45);
		}

		protected Boolean MathematicalFormula_impl_45()
		{
			return RecursionCall(MathematicalFormula_impl_0);
		}

		protected Boolean MathematicalFormula_impl_44()
		{
			String _literal_0 = "(";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_40()
		{
			return OneOrMore(MathematicalFormula_impl_41, "");
		}

		protected Boolean MathematicalFormula_impl_41()
		{
			String _classExpression_0 = "[0-9]";
			return CharacterClass(_classExpression_0, 5);
		}

		protected Boolean MathematicalFormula_impl_32()
		{
			String _nodeName_0 = "Symbol";
			return CapturingGroup(MathematicalFormula_impl_33, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_33()
		{
			return PrioritizedChoice(MathematicalFormula_impl_34, MathematicalFormula_impl_35);
		}

		protected Boolean MathematicalFormula_impl_35()
		{
			String _literal_0 = "-";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_34()
		{
			String _literal_0 = "+";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_4()
		{
			String _nodeName_0 = "Product";
			return CapturingGroup(MathematicalFormula_impl_5, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_5()
		{
			return Sequence(MathematicalFormula_impl_6, MathematicalFormula_impl_15);
		}

		protected Boolean MathematicalFormula_impl_15()
		{
			return ZeroOrMore(MathematicalFormula_impl_16, "");
		}

		protected Boolean MathematicalFormula_impl_16()
		{
			return Sequence(MathematicalFormula_impl_17, MathematicalFormula_impl_21);
		}

		protected Boolean MathematicalFormula_impl_21()
		{
			String _nodeName_0 = "Value";
			return CapturingGroup(MathematicalFormula_impl_22, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_22()
		{
			return PrioritizedChoice(MathematicalFormula_impl_23, MathematicalFormula_impl_25);
		}

		protected Boolean MathematicalFormula_impl_25()
		{
			return Sequence(MathematicalFormula_impl_26, MathematicalFormula_impl_29);
		}

		protected Boolean MathematicalFormula_impl_29()
		{
			String _literal_0 = ")";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_26()
		{
			return Sequence(MathematicalFormula_impl_27, MathematicalFormula_impl_28);
		}

		protected Boolean MathematicalFormula_impl_28()
		{
			return RecursionCall(MathematicalFormula_impl_0);
		}

		protected Boolean MathematicalFormula_impl_27()
		{
			String _literal_0 = "(";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_23()
		{
			return OneOrMore(MathematicalFormula_impl_24, "");
		}

		protected Boolean MathematicalFormula_impl_24()
		{
			String _classExpression_0 = "[0-9]";
			return CharacterClass(_classExpression_0, 5);
		}

		protected Boolean MathematicalFormula_impl_17()
		{
			String _nodeName_0 = "Symbol";
			return CapturingGroup(MathematicalFormula_impl_18, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_18()
		{
			return PrioritizedChoice(MathematicalFormula_impl_19, MathematicalFormula_impl_20);
		}

		protected Boolean MathematicalFormula_impl_20()
		{
			String _literal_0 = "/";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_19()
		{
			String _literal_0 = "*";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_6()
		{
			String _nodeName_0 = "Value";
			return CapturingGroup(MathematicalFormula_impl_7, _nodeName_0, false, false);
		}

		protected Boolean MathematicalFormula_impl_7()
		{
			return PrioritizedChoice(MathematicalFormula_impl_8, MathematicalFormula_impl_10);
		}

		protected Boolean MathematicalFormula_impl_10()
		{
			return Sequence(MathematicalFormula_impl_11, MathematicalFormula_impl_14);
		}

		protected Boolean MathematicalFormula_impl_14()
		{
			String _literal_0 = ")";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_11()
		{
			return Sequence(MathematicalFormula_impl_12, MathematicalFormula_impl_13);
		}

		protected Boolean MathematicalFormula_impl_13()
		{
			return RecursionCall(MathematicalFormula_impl_0);
		}

		protected Boolean MathematicalFormula_impl_12()
		{
			String _literal_0 = "(";
			return Literal(_literal_0, true);
		}

		protected Boolean MathematicalFormula_impl_8()
		{
			return OneOrMore(MathematicalFormula_impl_9, "");
		}

		protected Boolean MathematicalFormula_impl_9()
		{
			String _classExpression_0 = "[0-9]";
			return CharacterClass(_classExpression_0, 5);
		}
	}
}